src/osd/PrimaryLogPG.cc

```diff
@@ -9976,6 +9982,312 @@ struct C_Flush : public Context {
   }
 };
 
+struct C_SetDedupChunks : public Context {
+  PrimaryLogPGRef pg;
+  hobject_t oid;
+  epoch_t last_peering_reset;
+  ceph_tid_t tid;
+  uint64_t offset;
+  
+  C_SetDedupChunks(PrimaryLogPG *p, hobject_t o, epoch_t lpr, uint64_t offset)
+    : pg(p), oid(o), last_peering_reset(lpr),
+      tid(0), offset(offset)
+  {}
+  void finish(int r) override {
+    if (r == -ECANCELED)
+      return;
+    std::scoped_lock locker{*pg};
+    if (last_peering_reset != pg->get_last_peering_reset()) {
+      return;
+    }
+    pg->finish_set_dedup(oid, r, tid, offset);
+  }
+};
+
+
+int PrimaryLogPG::start_dedup(OpRequestRef op, ObjectContextRef obc)
+{
+  bufferlist bl;
+  uint64_t cur_off = 0;
+  const object_info_t& oi = obc->obs.oi;
+  const hobject_t& soid = oi.soid;
+
+  ceph_assert(obc->is_blocked());
+  if (oi.size == 0) {
+    // evicted 
+    return 0;
+  }
+  if (pool.info.get_fingerprint_type() == pg_pool_t::TYPE_FINGERPRINT_NONE) {
+    dout(0) << " fingerprint algorithm is not set " << dendl;
+    return 0;
+  } 
+
+  /*
+   * The operations to make dedup chunks are tracked by a ManifestOp.
+   * This op will be finished if all the operations are completed.
+   */
+  ManifestOpRef mop(std::make_shared<ManifestOp>(nullptr, 0));
+
+  while (cur_off < oi.size) {
+    // cdc
+    vector<pair<uint64_t, uint64_t>> chunks;
+    int r = do_cdc(oi, bl, chunks);
+    if (r < 0) {
+      return r;
+    }
+    if (!chunks.size()) {
+      break;
+    }
+
+    // get fingerprint 
+    for (auto p : chunks) {
+      bufferlist chunk;
+      chunk.substr_of(bl, p.first, p.second);
+      hobject_t target = get_fpoid_from_chunk(soid, chunk);
+
+      chunk_info_t chunk_info(0, p.second, target);
+      // chunks issued here are different with chunk_map newly generated
+      // because the same chunks in previous snap will not be issued
+      // So, we need two data structures; the first is the issued chunk list to track
+      // issued operations, and the second is the new chunk_map to update chunk_map after 
+      // all operations are finished
+      mop->new_chunk_map[p.first] = chunk_info;
+      // skip if the same content exits in prev snap at same offset
+      if (obc->ssc->snapset.clones.size()) {
```

Comment:
This portion below is unnecessarily wasteful.  You call get_prev_clone_obc for every chunk when clearly get_prev_clone_obc is a function of the object and not of the chunk.  You could easily declare set_chunk before the loop and then call calc_refs_to_inc_on_set after the loop on the accumulated set of references with cobc after calling get_prev_clone_obc a single time.  Note, calc_refs_to_inc_on_set *specifically* tolerates a null value passed in for both _g an _l, you shouldn't need to guard this call with an if.

Reply:
You haven't addressed this.  As I mentioned above, the outer while loop doesn't appear to be useful as bl will always contain the entire object after you call do_cdc.  
* Move get_fpoid_from_chunk into do_cdc and have it populate a mop->new_chunk_map directly.
* Change the type of mop->new_chunk_map to an object_manifest_t named mop->new_manifest.
* Once it is fully populated, pass mop->new_manifest to calc_refs_to_inc_on_set *once* and then use the resulting map to issue the refcounts.

I'd also really prefer that you not duplicate the logic in refcount_manifest.  At  least, I only want once place that encodes the "chunk_get_ref" names.  You should be able to modify refcount_manifest or one of the higher level helpers enough to be useful here.